fubectl.source

```diff
@@ -1,30 +1,54 @@
 # helper functions
 alias _inline_fzf="fzf --multi --ansi -i -1 --height=50% --reverse -0 --header-lines=1 --inline-info --border"
 alias _inline_fzf_nh="fzf --multi --ansi -i -1 --height=50% --reverse -0 --inline-info --border"
+_isNamespaced(){
+    if [[ "$1" == "" ]]; then
+        echo "please specify a Kubernetes object!"
+        exit 1
+    fi
+    # ask cluster about not namespaced resources
+    if [ $(kubectl api-resources --no-headers --namespaced=false | awk '{print $1,$2,$4}' | sed 's/ /\n/g' | grep --count -i "^$1$")  -gt 0 ]; then
```

Comment:
Why breaking this into seperate lines before doing a grep?

Reply:
To get exact match. Some thing like `pod` should not match to `PodSecurityPolicy`. 
I tried to get a regex for line with matching at 3 different positions. But didn't get a workable solution. This is the most understandable.

Reply:
So you produce the artificial exact match by replacing spaces into newlines and then match for newlines by using grep? I think you want to match for word breaks: `grep '\<pod\>' ...`. Or what is your boundary you want to match on?

Reply:
ok maybe my regex skill are not good enough. Feel free to change it. 
If i get this line:
```
clusters kubermatic.k8s.io Cluster
```
It should match case insensitive to the exact words e.g. `clusters` or `cluster`. But not if the search term is in the middle of the word, like `globalnetworkpolicies` should not match to `policies`.
For me this was the only way, where it works 100% perfect. So if you have a better solution, I looking forward how it looks like :man_shrugging:

Reply:
As given in my previous comment and now adapted to your code: `grep "\<$1\>".
Also regarding your `awk`: You are passing a `--namespaced=false` and still outputting the $4 which will always be `false`. Then `$1` and `$2` is the only thing you want to look into.

BUT!

The `kubectl api=resources` output is indeed a bit tricky and makes a simple `$1, $2` use-case invalid. There is no field separator like a tab character awk can use to distinct the columns. There is also no `-o jsonoutput`. So to make it correct (meaning you want to match on name and shortname only). The only valid thing to use is the header, as it denotes the character-length of the columns. 

So I propose:

```
_isNamespaced() {
  obj="$1"
  if [ -z "$obj" ]; then
    echo "please specify a Kubernetes object!"
    exit 1
  fi

  kubectl api-resources --namespaced=true \
    | awk '(idx){print substr($0, 0, idx) } (!idx){ idx=index($0, " APIGROUP")}' \
    | grep -q "\<${obj}\>"
}
```

So we are using the column-width here which we deduce from the header with awk. Then I removed the whole `if` construct as explained in my earlier comment.

Reply:
But an actual issue, which is addressed in #19 is the plural form. This _namespaced function would not match on singulars.

Reply:
The plural/singular problem can be fixed by adapting the grep with an optional s:
`grep -qE "\<${obj}s?\>"`

Reply:
The plural form is matching by the `$4` argument and `grep -i`. The `$4` value has the singular form just in Camel Casewriting. This more save, because the plural from cloud potential something without `s`, e.g. no plural s in `information`.
I will take a look at the above suggested solution and will reach you out again. Unfortunately I'm not a bash pro, so I try my best :yum:

Reply:
The problem with that is that in ~50% of the cases your $4 is also `false` because there is no filled 2nd column. If `KIND` is what should just be matched as well, the function changes to:
```
_isNamespaced() {
  obj="$1"
  if [ -z "$obj" ]; then
    echo "please specify a Kubernetes object!"
    exit 1
  fi  

  kubectl api-resources --namespaced=true \
    | awk '(apiidx){print substr($0, 0, apiidx),substr($0, kindidx) } (!apiidx){ apiidx=index($0, " APIGROUP");kindidx=index($0, " KIND")}' \
    | grep -iq "\<${obj}\>"
}
```